crates/apollo-compiler/src/validation/fragment.rs

```diff
@@ -180,54 +183,60 @@ pub fn validate_inline_fragment(
+    var_defs: Arc<Vec<hir::VariableDefinition>>,
 ) -> Vec<ApolloDiagnostic> {
     let mut diagnostics = Vec::new();
 
     diagnostics.extend(db.validate_directives(
         spread.directives().to_vec(),
         hir::DirectiveLocation::FragmentSpread,
+        var_defs.clone(),
     ));
 
-    if spread.fragment(db.upcast()).is_some() {
-        diagnostics
-            .extend(db.validate_fragment_selection(hir::FragmentSelection::FragmentSpread(spread)))
-    } else {
-        diagnostics.push(
-            ApolloDiagnostic::new(
-                db,
-                spread.loc().into(),
-                DiagnosticData::UndefinedFragment {
-                    name: spread.name().to_string(),
-                },
-            )
-            .labels(vec![Label::new(
-                spread.loc(),
-                format!("fragment `{}` is not defined", spread.name()),
-            )]),
-        );
+    match spread.fragment(db.upcast()) {
+        Some(def) => {
+            diagnostics.extend(
+                db.validate_fragment_selection(hir::FragmentSelection::FragmentSpread(spread)),
+            );
+            diagnostics.extend(db.validate_fragment_definition(def, var_defs));
+        }
+        None => {
+            diagnostics.push(
+                ApolloDiagnostic::new(
+                    db,
+                    spread.loc().into(),
+                    DiagnosticData::UndefinedFragment {
+                        name: spread.name().to_string(),
+                    },
+                )
+                .labels(vec![Label::new(
+                    spread.loc(),
+                    format!("fragment `{}` is not defined", spread.name()),
+                )]),
+            );
+        }
     }
 
     diagnostics
 }
 
-pub fn validate_fragment_definitions(
+pub fn validate_fragment_definition(
     db: &dyn ValidationDatabase,
-    file_id: FileId,
+    def: Arc<hir::FragmentDefinition>,
+    var_defs: Arc<Vec<hir::VariableDefinition>>,
 ) -> Vec<ApolloDiagnostic> {
     let mut diagnostics = Vec::new();
-    for def in db.fragments(file_id).values() {
-        diagnostics.extend(db.validate_directives(
-            def.directives().to_vec(),
-            hir::DirectiveLocation::FragmentDefinition,
-        ));
-        diagnostics.extend(
-            db.validate_fragment_type_condition(Some(def.type_condition().to_string()), def.loc()),
-        );
-        diagnostics.extend(db.validate_selection_set(def.selection_set().clone()));
-        diagnostics.extend(db.validate_fragment_used(Arc::clone(def), file_id));
+    diagnostics.extend(db.validate_directives(
+        def.directives().to_vec(),
+        hir::DirectiveLocation::FragmentDefinition,
+        var_defs.clone(),
+    ));
+    diagnostics.extend(
+        db.validate_fragment_type_condition(Some(def.type_condition().to_string()), def.loc()),
+    );
+    diagnostics.extend(db.validate_selection_set(def.selection_set().clone(), var_defs));
```

Comment:
The consequence of moving `db.validate_fragment_definition` to be run during selection validation is that we somehow end up missing some field validation inside a fragment definition's selection set! really confused on why that's happening

Reply:
fixed this bit!